hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSet.java

```diff
@@ -755,178 +697,99 @@ public static DataSet toSpliceLocatedRow(JavaRDD<Row> rdd, OperationContext cont
+            colNames = new String[descriptors.length];
+            int i = 0;
+            for (ResultColumnDescriptor rcd : export.getSourceResultColumnDescriptors()) {
+                colNames[i++] = rcd.getName();
             }
-            StructType tableSchema = DataTypes.createStructType(fields);
-
-            dataSchema = ExternalTableUtils.getDataSchema(dsp, tableSchema, partitionBy, location, "p");
-
-            if (dataSchema == null)
-                dataSchema = tableSchema;
-            // construct a DF using schema of data
-            Dataset<Row> insertDF = SpliceSpark.getSession().createDataFrame(
-                    rdd
-                            .map(new SparkSpliceFunctionWrapper<>(new CountWriteFunction(context)))
-                            .map(new LocatedRowToRowFunction()),
-                    dataSchema);
-
-            StructField[] dataSchemaFields = dataSchema.fields();
-            List<String> partitionByCols = new ArrayList();
-            for (int i = 0; i < partitionBy.length; i++) {
-                partitionByCols.add(dataSchemaFields[partitionBy[i]].name());
-            }
-            if (partitionBy.length > 0) {
-                List<Column> repartitionCols = new ArrayList();
-                for (int i = 0; i < partitionBy.length; i++) {
-                    repartitionCols.add(new Column(dataSchemaFields[partitionBy[i]].name()));
-                }
-                insertDF = insertDF.repartition(scala.collection.JavaConversions.asScalaBuffer(repartitionCols).toList());
-            }
-            insertDF.write().option(SPARK_COMPRESSION_OPTION,compression)
-                    .partitionBy(partitionByCols.toArray(new String[partitionByCols.size()]))
-                    .mode(SaveMode.Append).parquet(location);
-            ValueRow valueRow=new ValueRow(1);
-            valueRow.setColumn(1,new SQLLongint(context.getRecordsWritten()));
-            return new SparkDataSet<>(SpliceSpark.getContext().parallelize(Collections.singletonList(valueRow), 1));
-        } catch (Exception e) {
-            throw new RuntimeException(e);
+        } else {
+            throw new IllegalArgumentException("Unsupported operation type: " + context.getOperation());
+        }
+        StructField[] fields = new StructField[colNames.length];
+        for (int i=0 ; i<colNames.length ; i++){
+            fields[i] = dvds[i].getStructField(colNames[i]);
         }
+        StructType tableSchema = DataTypes.createStructType(fields);
+
+        // construct a DF using schema of data
+        Dataset<Row> insertDF = SpliceSpark.getSession().createDataFrame(
+                rdd
+                        .map(new SparkSpliceFunctionWrapper<>(new CountWriteFunction(context)))
+                        .map(new LocatedRowToRowFunction()),
+                tableSchema);
+
+        return new NativeSparkDataSet<>(insertDF, context).writeParquetFile(dsp, partitionBy, location, compression, context);
```

Comment:
It is true there's some redundancy on the tableSchema computation, it might be possible to eliminate it from NativeSparkDataset.writeParquetFile() (or remove the NativeSparkDataset creation). We need insertDF because Parquet files require a schema, and an RDD with a schema is a Dataframe (dataset)

Reply:
The duplicated logic shouldn't be too expensive to compute, since we are just creating a Spark schema. The only operation that had a duplicated expensive operation was WriteAvroFile, and that duplication can't be removed easily due to some issues with Avro and Date-types. What I did is to duplicate the logic to write the avro file in SparkDataset to avoid executing the potentially expensive operation twice ( ExternalTableUtils.getDataSchema )

Reply:
Thanks @dgomezferro I'm good with the change.